File: tests/radix-sort.test.mjs
import assert from "node:assert/strict";
import test from "node:test";
import { radixSort } from "../algorithms/sorting/radix-sort/radix-sort.mjs";

const cases = [
  { name: "empty input", input: [], expected: [] },
  { name: "single value", input: [7], expected: [7] },
  { name: "already sorted", input: [1, 2, 3, 4, 5], expected: [1, 2, 3, 4, 5] },
  { name: "reverse order", input: [500, 40, 3, 2, 1], expected: [1, 2, 3, 40, 500] },
  { name: "duplicates", input: [21, 11, 21, 12], expected: [11, 12, 21, 21] },
  { name: "negative values", input: [-12, 5, -8, 0], expected: [-12, -8, 0, 5] },
];

for (const { name, input, expected } of cases) {
  test(`Radix Sort handles ${name}`, () => {
    const original = [...input];
    assert.deepEqual(radixSort(input), expected);
    assert.deepEqual(input, original, "input should not be mutated");
  });
}

test("Radix Sort agrees with the platform sort oracle for deterministic random inputs", () => {
  let seed = 109;
  const next = () => {
    seed = (seed * 48271) % 2147483647;
    return seed;
  };

  for (let round = 0; round < 30; round += 1) {
    const input = Array.from({ length: 12 }, () => (next() % 1001) - 500);
    assert.deepEqual(radixSort(input), [...input].sort((a, b) => a - b));
  }
});

File: algorithms/sorting/radix-sort/RadixSortTest.java
import java.util.Arrays;
import java.util.Random;

public final class RadixSortTest {
    private RadixSortTest() {
    }

    private static void verify(String name, int[] input, int[] expected) {
        int[] original = Arrays.copyOf(input, input.length);
        int[] actual = RadixSort.radixSort(input);
        if (!Arrays.equals(actual, expected)) throw new AssertionError(name + " failed: " + Arrays.toString(actual));
        if (!Arrays.equals(input, original)) throw new AssertionError(name + " mutated its input");
    }

    public static void main(String[] arguments) {
        verify("empty input", new int[] {}, new int[] {});
        verify("single value", new int[] {7}, new int[] {7});
        verify("already sorted", new int[] {1, 2, 3, 4, 5}, new int[] {1, 2, 3, 4, 5});
        verify("reverse order", new int[] {500, 40, 3, 2, 1}, new int[] {1, 2, 3, 40, 500});
        verify("duplicates", new int[] {21, 11, 21, 12}, new int[] {11, 12, 21, 21});
        verify("negative values", new int[] {-12, 5, -8, 0}, new int[] {-12, -8, 0, 5});

        Random random = new Random(109);
        for (int round = 0; round < 30; round++) {
            int[] input = new int[12];
            for (int index = 0; index < input.length; index++) input[index] = random.nextInt(1001) - 500;
            int[] expected = Arrays.copyOf(input, input.length);
            Arrays.sort(expected);
            verify("random input " + round, input, expected);
        }

        System.out.println("Java Radix Sort tests passed");
    }
}
